### ISSUE TITLE
"// pico_cmake_set" not seen in included header files.

### ISSUE THREAD

**Issue by Memotech-Bill** [raspberrypi/pico-sdk#2112]
If a board header file includes other header files, then CMake does not see any **pico_cmake_set** directives in the secondary header file.

This particularly affects the **vgaboard.h** header file, which includes **pico.h**. None of the pico settings are seen by CMake. This does not matter in this case as all the settings are default values anyway.

However, it is not possible to simply change **pico.h** to **pico2.h** for a VGA board with pico2 installed, as CMake will still think it is compiling for RP2040 rather than RP2350.

**Comment by lurch** [comment]
FWIW, https://github.com/raspberrypi/pico-sdk/blob/develop/tools/check_board_header.py _does_ look at `#inclue` files :laughing:

**Comment by lurch** [comment]
I think the search for `#include` lines would need to be recursive though, since you might well have nested `#include`s?

**Comment by will-v-pi** [comment]
That's true - in which case maybe add the parsing into the `while(HEADER_FILE_CONTENTS)` loop, and append to `HEADER_FILE_CONTENTS` when you see an include directive. This would also have the effect of processing the `cmake_set_default`s from the included header after the new header, which I think is the preferred way round (so you can override them in the new header)?

**Comment by lurch** [comment]
I guess _appending_ to `HEADER_FILE_CONTENTS` (rather than doing an "in-place insert") might theoretically cause a problem if you had something like:

`file1.h`
```
// pico_cmake_set VAR1=VAL1
// pico_cmake_set VAR2=VAL2
#include "file2.h"
// pico_cmake_set VAR3=VAL3
```

`file2.h`
```
// pico_cmake_set VAR2=VAL4
// pico_cmake_set VAR3=VAL5
// pico_cmake_set VAR4=VAL6
```

I think you'd "expect" `VAR1=VAL1, VAR2=VAL4, VAR3=VAL3, VAR4=VAL6` but if you only appended to `HEADER_FILE_CONTENTS` when encountering an `#include` then you'd actually end up with `VAR1=VAL1, VAR2=VAL4, VAR3=VAL5, VAR4=VAL6` ?
But this may not _actually_ be a problem, and doing a simple append-only approach might be "good enough" for the cases where `pico_cmake_set` is actually used?

**Comment by will-v-pi** [comment]
That's very true (with the opposite expected for `cmake_set_default`) - so it would be better to pre-pend the new file to `HEADER_FILE_CONTENTS` (the same as inserting in-place as we're popping from the front), which should be simple enough.

I'll steal your regex from `check_board_header.py` and whip up a PR

**Comment by Memotech-Bill** [comment]
I have wondered whether BOARDS could optionally be a list, rather than a single filename.

In that way, a pico2_w on a VGA board might be defined as:

````
BOARD=pico2_w,vgaboard
````

That would save having to have multiple definitions for any of the boards which will accept any flavour of pico.

**Comment by lurch** [comment]
> I have wondered whether BOARD could optionally be a list, rather than a single filename.

Ugh no. That might lead to people thinking that `BOARD=pico2,pico_w` is the same thing as `BOARD=pico2_w`.

I guess some concept of `CARRIER_BOARD` might be useful (see e.g. #696 and #837 ), but I guess that's a much bigger change.

**Comment by lurch** [comment]
@Memotech-Bill Would you like to give #2113 a try please, and see if it behaves as you expect?

**Comment by Memotech-Bill** [comment]
> we intend to make a PICO_CARRIER_BOARD (or CARRIED board) to cover those, which could have helped, but will think about recursion too... In this case probably a good idea

This should perhaps be PICO_ADDON_BOARDS, and should probably have the capability to be a list. Unlike the RPi, Pico addon boards do not have configuration EEPROMs, so it is straightforward to stack multiple addon boards.

I can't call them HATs as the usually go underneath the Pico, not on top.

**Comment by dpgeorge** [comment]
We've hit this problem in MicroPython, trying to extend the definition of a pico-sdk provided board header (so, include the base header and add a few defines).  The derived board header fails to pick up the `pico_cmake_set`/etc directives in the base board header.

At the moment the workaround we have is to copy the `pico_cmake_set` directives manually to the derived board header.  But that's pretty ugly.

IMO it'd be much simpler to have a board defined at the cmake level, eg `pico.cmake`, instead of at the C level and needing to duplicate config lines, eg for flash size with both a `#define` and `pico_cmake_set`.  And that cmake config file sets all the relevant values at the cmake level, with the build system then propagating those cmake variables into C defines via `target_compile_definitions`.  That's a big change though...

**Comment by lurch** [comment]
@will-v-pi What happened to #2113 ?

**Comment by will-v-pi** [comment]
> [@will-v-pi](https://github.com/will-v-pi) What happened to [#2113](https://github.com/raspberrypi/pico-sdk/pull/2113) ?

Looks like I accidentally closed it - re-opened as #2968

**Comment by lurch** [comment]
Fixed by #2968
